### ISSUE TITLE
Conditional string templates don't work

### ISSUE THREAD

**Issue by arteam** [jdbi/jdbi#413]
Consider a simple StringTemplate example:

```java
public class TestConditionalStringTemplateLocator {

    @Rule
    public H2DatabaseRule db = new H2DatabaseRule().withPlugin(new SqlObjectPlugin());

    @Before
    public void setUp() throws Exception {
        Handle handle = db.getSharedHandle();
        handle.insert("insert into something (id, name) values (1, 'Martin')");
        handle.insert("insert into something (id, name) values (2, 'Joe')");
        handle.insert("insert into something (id, name) values (3, 'David')");
    }

    @Test
    public void testFindAndSortByName() {
        List<Integer> ids = db.getSharedHandle().attach(Dao.class).find(true, "name");
        System.out.println(ids);
    }

    @Test
    public void testFindWithoutSorting() {
        List<Integer> ids = db.getSharedHandle().attach(Dao.class).find(false, "");
        System.out.println(ids);
    }

    @UseStringTemplateSqlLocator
    interface Dao {
        @SqlQuery
        List<Integer> find(@Define("sort") boolean sort, @Define("sortBy") String sortBy);
    }
}
```

```
group Dao;

find(sort,sortBy) ::= << select id from something <if(sort)> order by <sortBy> <endif> >>
```

On the head of the `jdbi3` branch this test case returns the following exception:
```
org.jdbi.v3.core.exception.UnableToCreateStatementException: Exception parsing for named parameter replacement [statement:" select id from something <if(sortBy)> order by <sortBy> <endif> ", rewritten:"null", arguments:{ positional:{}, named:{}, finder:[]}]

	at org.jdbi.v3.core.rewriter.ColonPrefixStatementRewriter.rewrite(ColonPrefixStatementRewriter.java:62)
	at org.jdbi.v3.core.SqlStatement.internalExecute(SqlStatement.java:1236)
	at org.jdbi.v3.core.Query.iterator(Query.java:69)
	at org.jdbi.v3.core.ResultBearing.stream(ResultBearing.java:83)
	at org.jdbi.v3.core.ResultBearing.collect(ResultBearing.java:137)
	at org.jdbi.v3.sqlobject.ResultReturnThing$DefaultResultReturnThing.result(ResultReturnThing.java:118)
	at org.jdbi.v3.sqlobject.ResultReturnThing.map(ResultReturnThing.java:48)
	at org.jdbi.v3.sqlobject.QueryHandler.invoke(QueryHandler.java:42)
	at org.jdbi.v3.sqlobject.SqlObjectFactory.lambda$createMethodInterceptor$9(SqlObjectFactory.java:167)
	at org.jdbi.v3.stringtemplate.TestConditionalStringTemplateLocator$Dao$$EnhancerByCGLIB$$cd04f13c.find(<generated>)
	at org.jdbi.v3.stringtemplate.TestConditionalStringTemplateLocator.testFind(TestConditionalStringTemplateLocator.java:27)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:119)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
Caused by: java.lang.IllegalArgumentException: MismatchedTokenException(40!=62)
	at org.jdbi.v3.core.internal.lexer.DefineStatementLexer.reportError(DefineStatementLexer.java:30)
	at org.antlr.runtime.Lexer.nextToken(Lexer.java:103)
	at org.jdbi.v3.core.rewriter.DefinedAttributeRewriter.rewriteDefines(DefinedAttributeRewriter.java:57)
	at org.jdbi.v3.core.rewriter.ColonPrefixStatementRewriter.rewrite(ColonPrefixStatementRewriter.java:59)
	... 39 more
Caused by: MismatchedTokenException(40!=62)
	at org.antlr.runtime.Lexer.match(Lexer.java:197)
	at org.jdbi.v3.core.internal.lexer.DefineStatementLexer.mGT(DefineStatementLexer.java:149)
	at org.jdbi.v3.core.internal.lexer.DefineStatementLexer.mDEFINE(DefineStatementLexer.java:543)
	at org.jdbi.v3.core.internal.lexer.DefineStatementLexer.mTokens(DefineStatementLexer.java:1116)
	at org.antlr.runtime.Lexer.nextToken(Lexer.java:89)
	... 41 more
```

This test case works with `jdbi3-3.0.0-alpha0`, so it looks like that the refactoring in #391 broke this functionality. 

At first glance, it looks like that we still need to pass a `StatementContext` to locators,
so `StringTemplateSqlLocator` can have access to attributes defined by the user and reevaluate string templates.

**Comment by arteam** [comment]
An alternate approach is to introduce a new annotation (for example, `@Argument`) in the `string-template` module to differentiate between `StringTemplate` attributes and "JDBI" attributes handled by `DefinedAttributeRewriter`. `StringTemplateSqlLocator` can handle such attributes internally without a statement context.

**Comment by qualidafial** [comment]
Yep, this was an oversight. I was going to add a StringTemplateStatementRewriter but spaced it off. I thought I submitted an issue for it, but it may still be in an open tab waiting to click submit.

**Comment by qualidafial** [comment]
There's another thing I realized: by separating the string template locating and rewriting, we might have lost the ability for one string template in a group to inline another template in the group by reference.

**Comment by qualidafial** [comment]
I think the following should fix this issue as reported:

* Add `StringTemplateStatementRewriter`
* Add `@UseStringTemplateStatementRewriter`
* Make `@UseStringTemplateSqlLocator` implicitly imply the other annotation, since that kinda makes sense. So locate with StringTemplate implies rewrite also, But rewrite by itself means that the SQL method annotation value is used as the raw string template.

**Comment by arteam** [comment]
Sounds like a plan.

**Comment by qualidafial** [comment]
Resolved in #421
